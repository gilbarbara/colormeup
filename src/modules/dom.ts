import { Types } from '@gilbarbara/components';
import is from 'is-lite';

import { breakpoints } from '~/config';

/**
 * Get screen size breakpoint
 */
export function getBreakpoint(): Types.Breakpoint {
  const windowWidth = window.innerWidth;
  let breakpoint = 'xxs';

  /* istanbul ignore next */
  if (windowWidth >= 1920) {
    breakpoint = 'xxxl';
  } else if (windowWidth >= 1440) {
    breakpoint = 'xxl';
  } else if (windowWidth >= 1280) {
    breakpoint = 'xl';
  } else if (windowWidth >= 1024) {
    breakpoint = 'lg';
  } else if (windowWidth >= 768) {
    breakpoint = 'md';
  } else if (windowWidth >= 400) {
    breakpoint = 'sm';
  } else if (windowWidth >= 375) {
    breakpoint = 'xs';
  }

  return breakpoint as Types.Breakpoint;
}

/**
 * Return true for any screen size below a given breakpoint
 */
export function maxScreenSize(breakpoint: Types.Breakpoint | number = 'md') {
  return !minScreenSize(breakpoint);
}

/**
 * Return true for any screen size above a given breakpoint
 */
export function minScreenSize(breakpoint: Types.Breakpoint | number = 'md') {
  const width = window.innerWidth;

  if (is.number(breakpoint)) {
    return width >= breakpoint;
  }

  return width >= breakpoints[breakpoint];
}
