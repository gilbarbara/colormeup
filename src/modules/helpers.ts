import { now } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';
import { REHYDRATE } from 'redux-persist';

import { modes } from '~/config';

import { ModeType, RootState } from '~/types';

export function actionBody<T = any>(payload: T) {
  return { payload };
}

export function getColorModes() {
  return Object.entries(modes).reduce<Array<{ name: string; types: ModeType[] }>>(
    (acc, [name, mode]) => {
      const keys = Object.keys(mode);

      acc.push({
        name,
        types: keys.map(d => ({
          ...mode[d],
          key: d,
          slug: mode[d].name.toLowerCase(),
        })),
      });

      return acc;
    },
    [],
  );
}

/**
 * Check if cache is valid
 */
export function hasValidCache(lastUpdated: number, max = 10): boolean {
  if (!navigator.onLine) {
    return true;
  }

  return lastUpdated + max * 60 > now();
}

/**
 * Test if the input is a valid number
 */
export function isNumber(input: unknown): input is number {
  return typeof input === 'number' && !Number.isNaN(input);
}

export function parseColorInput(input: string) {
  return `#${input.replace(/^#/, '')}`;
}

export const rehydrateAction = createAction<RootState>(REHYDRATE);
