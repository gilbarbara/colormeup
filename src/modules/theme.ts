import { mergeTheme } from '@gilbarbara/components';

export const theme = mergeTheme({
  colors: {
    primary: '#000',
  },
});

export const easing = 'cubic-bezier(0.35, 0.01, 0.77, 0.34);';
