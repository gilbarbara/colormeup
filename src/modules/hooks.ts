import { useCallback, useMemo, useRef } from 'react';
import { Selector, shallowEqual, useDispatch, useSelector } from 'react-redux';
import { useDeepCompareEffect } from 'react-use';
import { ASYNC_STATUS } from '@gilbarbara/helpers';
import { AsyncFlowWithDataAndCache } from '@gilbarbara/types';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';
import is from 'is-lite';

import { hasValidCache } from '~/modules/helpers';

import { RootState } from '~/types';

export const useAppDispatch = () => useDispatch();

export function useAppSelector<TReturn>(selector: (state: RootState) => TReturn) {
  return useSelector(selector, shallowEqual);
}

export function useQuery<T>(
  selector: Selector<RootState, AsyncFlowWithDataAndCache<T>>,
  action: ActionCreatorWithPayload<any>,
  ...parameters: any[]
) {
  const dispatch = useAppDispatch();
  const { data, message, status, updatedAt } = useAppSelector(selector);
  const hasData = useMemo(() => (is.array(data) ? !!data.length : !!data), [data]);
  const actionParameters = useRef(parameters);

  useDeepCompareEffect(() => {
    actionParameters.current = parameters;

    if ((!hasData && status !== ASYNC_STATUS.SUCCESS) || !hasValidCache(updatedAt)) {
      dispatch(action(parameters));
    }
  }, [action, dispatch, hasData, parameters, updatedAt]);

  const refresh = useCallback(() => {
    dispatch(action(actionParameters.current));
  }, [action, dispatch]);

  return {
    data,
    isLoading: status === ASYNC_STATUS.PENDING || (status === ASYNC_STATUS.IDLE && !hasData),
    message,
    refresh,
    status,
  };
}
