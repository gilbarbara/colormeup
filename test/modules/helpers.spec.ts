import { getColorModes, isNumber } from '~/modules/helpers';

describe('modules/helpers', () => {
  describe('getColorModes', () => {
    it('should return an array with the color modes', () => {
      expect(getColorModes()).toMatchSnapshot();
    });
  });

  describe('isNumber', () => {
    it('should return a boolean', () => {
      expect(isNumber(1)).toBe(true);
      expect(isNumber('1')).toBe(false);
      expect(isNumber('a')).toBe(false);
    });
  });
});
