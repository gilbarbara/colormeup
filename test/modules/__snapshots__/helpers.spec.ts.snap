// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modules/helpers > getColorModes > should return an array with the color modes 1`] = `
[
  {
    "name": "hsl",
    "types": [
      {
        "key": "h",
        "max": 360,
        "name": "Hue",
        "slug": "hue",
      },
      {
        "key": "s",
        "max": 100,
        "name": "Saturation",
        "slug": "saturation",
      },
      {
        "key": "l",
        "max": 100,
        "name": "Lightness",
        "slug": "lightness",
      },
    ],
  },
  {
    "name": "rgb",
    "types": [
      {
        "key": "r",
        "max": 255,
        "name": "Red",
        "slug": "red",
      },
      {
        "key": "g",
        "max": 255,
        "name": "Green",
        "slug": "green",
      },
      {
        "key": "b",
        "max": 255,
        "name": "Blue",
        "slug": "blue",
      },
    ],
  },
]
`;
